wasm-runtime/polkadot/src/support/proposal.rs

```diff
@@ -0,0 +1,133 @@
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Proposal: This describes a combination of a function ID and data that can be used to call into
+//! an internal function.
+
+use runtime_support::prelude::*;
+use runtime_support::mem;
+use slicable::Slicable;
+use joiner::Joiner;
+use streamreader::StreamReader;
+use runtime::{system, governance, staking, session};
+
+/// Internal functions that can be dispatched to.
+#[cfg_attr(test, derive(PartialEq, Debug))]
+#[derive(Clone, Copy)]
+pub enum InternalFunction {
+	SystemSetCode,
+	StakingSetSessionsPerEra,
+	StakingSetBondingDuration,
+	StakingSetValidatorCount,
+	GovernanceSetApprovalPpmRequired,
+	SessionSetLength,
+}
+
+impl InternalFunction {
+	/// Derive `Some` value from a `u8`, or `None` if it's invalid.
+	pub fn from_u8(value: u8) -> Option<InternalFunction> {
+		match value {
+			x if x == InternalFunction::SystemSetCode as u8 => Some(InternalFunction::SystemSetCode),
+			x if x == InternalFunction::StakingSetSessionsPerEra as u8 => Some(InternalFunction::StakingSetSessionsPerEra),
+			x if x == InternalFunction::StakingSetBondingDuration as u8 => Some(InternalFunction::StakingSetBondingDuration),
+			x if x == InternalFunction::StakingSetValidatorCount as u8 => Some(InternalFunction::StakingSetValidatorCount),
+			x if x == InternalFunction::GovernanceSetApprovalPpmRequired as u8 => Some(InternalFunction::GovernanceSetApprovalPpmRequired),
+			x if x == InternalFunction::SessionSetLength as u8 => Some(InternalFunction::SessionSetLength),
+			_ => None,
+		}
+	}
+}
+
+/// An internal function.
+#[cfg_attr(test, derive(PartialEq, Debug))]
+pub struct Proposal {
+	/// The priviledged function to call.
+	pub function: InternalFunction,
+	/// The serialised data to call it with.
+	pub input_data: Vec<u8>,
+}
+
+impl Slicable for Proposal {
+	fn set_as_slice<F: Fn(&mut[u8], usize) -> bool>(fill_slice: &F) -> Option<Self> {
+		Some(Proposal {
+			function: InternalFunction::from_u8(Slicable::set_as_slice(fill_slice)?)?,
+			input_data: Slicable::set_as_slice(&|s, o| fill_slice(s, o + 1))?,
+		})
+	}
+
+	fn to_vec(&self) -> Vec<u8> {
+		Vec::new()
+			.join(&(self.function as u8))
+			.join(&self.input_data)
+	}
+
+	fn size_of(data: &[u8]) -> Option<usize> {
+		let first_part = mem::size_of::<u8>();
+		let second_part = <Vec<u8>>::size_of(&data[first_part..])?;
+		Some(first_part + second_part)
+	}
+}
+
+impl Proposal {
+	pub fn enact(&self) {
+		let mut params = StreamReader::new(&self.input_data);
+		match self.function {
+			InternalFunction::SystemSetCode => {
+				let code: Vec<u8> = params.read().unwrap();
```

Comment:
panic proofs are missing in this match

Reply:
no need as this is runtime code. if anything goes wrong, the block is invalid and it'll be by the panic handler.